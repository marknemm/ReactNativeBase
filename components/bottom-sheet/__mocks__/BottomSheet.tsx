import type { BottomSheetProps } from '@components/bottom-sheet/BottomSheet.interfaces';

const MockBottomSheet = jest.fn().mockImplementation((props: BottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/bottom-sheet/BottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockBottomSheet;
