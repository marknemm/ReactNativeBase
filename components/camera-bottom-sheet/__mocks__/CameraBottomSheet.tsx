import type { CameraBottomSheetProps } from '@components/camera-bottom-sheet/CameraBottomSheet.interfaces';

const MockCameraBottomSheet = jest.fn().mockImplementation((props: CameraBottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/camera-bottom-sheet/CameraBottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockCameraBottomSheet;
