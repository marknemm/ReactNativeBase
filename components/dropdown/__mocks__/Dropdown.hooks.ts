/**
 * Mocks the useDropdownItems hook.
 */
export const useDropdownItems = jest.fn().mockImplementation((data) => data);
