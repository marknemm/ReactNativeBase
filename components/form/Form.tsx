import { FormProvider } from 'react-hook-form';
import { View } from 'react-native';
import type { FormProps } from './Form.interface';

/**
 * A component that wraps a form and provides form context to its children.
 *
 * @param props The component {@link FormProps}.
 * @returns The {@link Form} component.
 */
const Form: React.FC<FormProps> = ({ children, form, style, ...restViewProps }) => (
  <FormProvider {...form}>
    <View
      style={style}
      testID="rnb-form-view"
      {...restViewProps}
    >
      {children}
    </View>
  </FormProvider>
);

export type * from './Form.interface';
export default Form;
