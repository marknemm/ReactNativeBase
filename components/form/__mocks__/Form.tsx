import type { FormProps } from '@components/form/Form.interface';

const FormMock = jest.fn().mockImplementation((props: FormProps) => {
  const ActualComponent = jest.requireActual('@components/form/Form').default;
  return <ActualComponent {...props} />;
});

export default FormMock;
