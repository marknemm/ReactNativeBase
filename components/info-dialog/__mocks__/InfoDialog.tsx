import type { InfoDialogProps } from '@components/info-dialog/InfoDialog.interfaces';

const MockInfoDialog = jest.fn().mockImplementation((props: InfoDialogProps) => {
  const ActualComponent = jest.requireActual('@components/info-dialog/InfoDialog').default;
  return <ActualComponent {...props} />;
});

export default MockInfoDialog;
