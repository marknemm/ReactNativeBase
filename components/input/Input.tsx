import { useFormControl, useFormErrorMessage, useValidationRules } from '@hooks/form-hooks';
import { useCallbacks } from '@hooks/state-hooks';
import { Input as RneInput, useTheme } from '@rneui/themed';
import { forwardRef, useEffect, useRef, useState } from 'react';
import { Controller } from 'react-hook-form';
import { useMaskedInputProps } from 'react-native-mask-input';
import type { InputFC, InputProps, InputRefType } from './Input.interfaces';
import { useStyles } from './Input.styles';

/**
 * Component for generic user text input.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link Input} component.
 */
const Input: InputFC = forwardRef((props, ref) => {
  const { label, name, onBlur, onChangeText } = props;

  // Derive entities related to an input controlled by react-hook-form
  const control = useFormControl(props);
  const rules = useValidationRules(props, label?.toString());
  const errorMessage = useFormErrorMessage(rules, props);

  return control
    ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange: onChangeForm, onBlur: onBlurForm, value } }) => (
          <InputControlled
            {...props}
            errorMessage={errorMessage}
            onBlur={(event) => {
              onBlurForm();
              onBlur?.(event);
            }}
            onChangeText={(text) => {
              onChangeForm(text);
              onChangeText?.(text);
            }}
            ref={ref}
            value={value}
          />
        )}
        rules={rules}
      />
    )
    : <InputControlled {...props} ref={ref} />;
});

/**
 * Controlled {@link Input} component.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link InputControlled} component.
 */
const InputControlled: InputFC = forwardRef(({
  containerStyle,
  inputContainerStyle,
  inputStyle,
  labelStyle,
  mask,
  maxLength,
  maxLengthLimitTyping,
  onChangeText,
  style,
  value,
  ...inputProps
}, ref) => {
  const styles = useStyles({ containerStyle, inputContainerStyle, inputStyle, labelStyle, style });
  const { theme } = useTheme();
  const [uiValue, setUiValue] = useState('');
  const inputRef = useRef<InputRefType>();
  ref ??= inputRef;

  const maskedInputProps = useMaskedInputProps({ mask, onChangeText, value });
  if (mask) {
    onChangeText = maskedInputProps.onChangeText;
    value = maskedInputProps.value;
  }

  const maxLengthNum = (typeof maxLength === 'number')
    ? maxLength
    : maxLength?.value ?? undefined;

  useEffect(() => {
    // Sync value prop with input (UI) value. Done manually to prevent change on each keystroke and prevent lag.
    if (value !== undefined && value !== uiValue) {
      const textInput = (ref as React.MutableRefObject<InputRefType>).current;
      textInput?.setNativeProps({ text: value });
    }
  }, [ref, value, uiValue]);

  return (
    <RneInput
      keyboardAppearance={theme.mode}
      placeholderTextColor={theme.colors.placeholder}
      {...inputProps}
      {...(mask ? maskedInputProps : undefined)}
      containerStyle={styles.container}
      inputContainerStyle={styles.inputContainer}
      inputStyle={styles.input}
      labelStyle={styles.label}
      maxLength={maxLengthLimitTyping ? maxLengthNum : undefined}
      onChangeText={useCallbacks(onChangeText, setUiValue)}
      ref={ref}
      style={styles.style}
      value={undefined} // Do not update value prop directly to prevent input lag (see useEffect above).
    />
  );
});

export type * from './Input.interfaces';
export default Input;
