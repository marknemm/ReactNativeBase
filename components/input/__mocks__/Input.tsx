import type { InputProps } from '@components/input/Input.interfaces';

const MockInput = jest.fn().mockImplementation((props: InputProps) => {
  const ActualComponent = jest.requireActual('@components/input/Input').default;
  return <ActualComponent {...props} />;
});

export default MockInput;
