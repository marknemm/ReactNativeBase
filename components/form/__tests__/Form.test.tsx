import Form from '@components/form/Form';
import Input from '@components/input/Input';
import { genMockForm, type DefaultMockFormT } from '@hooks/__mocks__/form-hooks';
import { Text } from '@rneui/themed';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, screen } from '@testing-library/react-native';
import { type UseFormReturn } from 'react-hook-form';

describe('<Form />', () => {
  let form: UseFormReturn<DefaultMockFormT>;

  beforeEach(() => {
    form = genMockForm();
  });

  describe('custom styles', () => {
    it('applies custom styles', () => {
      const style = { backgroundColor: 'black' };
      render(
        <Form
          form={form}
          style={style}
        >
          <Text>Form</Text>
        </Form>,
        { wrapper: AppProvider }
      );

      const formView = screen.getByTestId('rnb-form-view');
      expect(formView).toHaveStyle(style);
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <Form form={form} />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with children', () => {
      render(
        <Form form={form}>
          <Input name="field1" />
          <Input name="field2" />
        </Form>,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
