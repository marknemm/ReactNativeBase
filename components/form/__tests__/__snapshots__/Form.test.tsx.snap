// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> snapshots renders correctly 1`] = `
<View
  testID="rnb-form-view"
/>
`;

exports[`<Form /> snapshots renders correctly with children 1`] = `
<View
  testID="rnb-form-view"
>
  <View
    style={
      {
        "marginBottom": 6.285714285714286,
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
    testID="RNE__Input__view-wrapper"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "height": 98.9730612244898,
          "transform": [
            {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        editable={true}
        keyboardAppearance="light"
        name="field1"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholderTextColor="#86939e"
        style={
          {
            "color": "#242424",
            "flex": 1,
            "fontSize": 23.57142857142857,
            "height": 98.9730612244898,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
    <Text
      style={
        {
          "color": "#ff3b30",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
  <View
    style={
      {
        "marginBottom": 6.285714285714286,
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
    testID="RNE__Input__view-wrapper"
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "#bbb",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "height": 98.9730612244898,
          "transform": [
            {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        editable={true}
        keyboardAppearance="light"
        name="field2"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholderTextColor="#86939e"
        style={
          {
            "color": "#242424",
            "flex": 1,
            "fontSize": 23.57142857142857,
            "height": 98.9730612244898,
            "minHeight": 40,
          }
        }
        testID="RNE__Input__text-input"
        underlineColorAndroid="transparent"
      />
    </View>
    <Text
      style={
        {
          "color": "#ff3b30",
          "fontSize": 12,
          "margin": 5,
        }
      }
    />
  </View>
</View>
`;
